docs/performance-metrics.md

```diff
@@ -0,0 +1,98 @@
+
+## Common Bottlenecks
+
+### Memory
+
+ETCD uses bbolt, which stores everything as a single memory mapped file. In OpenShift this is up to 8 GiB big, this file is organized with a B+Tree and chunked into discrete 4k pages which can be flagged in-use or not. On top of that, several in-memory caches/indices on both ETCD and API server are being used to facilitate fast multi-version access and label/field selection respectively.
+
+Due to the indices and caches employed, it is fairly easy to exhaust memory on the control plane even though the actual used size of the etcd database file is tiny. One particular scenario where is this common is when listing all pods in a large cluster or listing many large configmaps/secrets.
+
+Check these metrics:
+
+> process_resident_memory_bytes{job="etcd"}
+> 
+> process_resident_memory_bytes{job="apiserver"}
+
+It is perfectly normal for etcd to use a multiple of its "in use DB size" as resident memory.
+
+Try to look through the audit log and find service accounts that might be listing pods or configmaps across namespaces.
+Another frequent offender are events from chatty operators, which can be found in the audit log or directly from [etcd using KCS6970015](https://access.redhat.com/solutions/6970015).
+
+#### Defrag and Compaction
+
+One way to reduce memory usage is to defrag bbolt and compact the database to the latest revision. It is important to run the defrag after the compaction, so one would execute:
+
+> etctctl compact $(etctctl get / -w json | jq .header.revision)
+> 
+> etctctl defrag
+
+In more recent OpenShift versions this is done automatically -  compacting from the API server and defrag from the etcd-operator.
+
+Sometimes etcd is quite reluctant to free memory, due to the laziness of the Golang GC. So it is perfectly normal for the resident memory to stay up until memory pressure forces it to go down again.
+
+### CPU
+
+While ETCD is not very heavy user of CPU in general, similar to memory, ETCD is competing for CPU usage with other components on the same hosts.
+
+This query will give you our most offending candidates (API server, OVN and ETCD) in a single chart. 
+
+> sum(node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate{namespace=~"openshift-etcd|openshift-ovn-kubernetes|openshift-apiserver"}) by (namespace)
+
+Especially when dealing with OVN, you might want to drill deeper into what component is using the CPU:
+
+> rate(container_cpu_usage_seconds_total{namespace=~"openshift-etcd|openshift-ovn-kubernetes", container=~"etcd|ovnkube-master|nbdb|northd|sbdb"}[2m])
+
+Keep in mind that this is not exhaustive, there might be something else that runs on the host that consumes lots of CPU and thus starves etcd. It also makes sense to look at the whole picture with:
+
+> sum(node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate{cluster="", node=~"!!! NODE NAME !!!"}) by (pod)
+ 
+
+### Disk Latency
```

Comment:
@tjungblu Have you ever had any experience with looking ata node's IOwait % metric?

I have this monstrosity saved from when I saw Sam using it long ago to see if some leader elections that we were seeing were correlated with IOwait spikes which indicated disk throttling.

```
(sum(irate(node_cpu_seconds_total {mode="iowait"} [2m])) without (cpu)) / count(node_cpu_seconds_total) without (cpu) * 100
AND on (instance) label_replace( kube_node_role{role="master"}, "instance", "$1", "node", "(.+)" )
```
Also in the KCS article here: https://access.redhat.com/solutions/5489721 which mentions it should be below 4%.

Again not sure how useful that is since being throttled in IOPS or a slow disk would likely manifest itself in the other metrics below as well. Just thought I'd mention it in case it seems relevant.

Reply:
ohh this is amazing. I was actually looking for that the other day, will add this in here too.